jaxline/utils.py

```diff
@@ -571,6 +571,15 @@ def save(self, ckpt_series: str) -> None:
     active_state = self.get_experiment_state(ckpt_series)
     id_ = 0 if not series.history else series.history[-1].id + 1
     snapshot = copy.copy(active_state)
+    for sk, sv in snapshot.items():
+      # Duck-typing for "is this a Jaxline Experiment class?".
+      if hasattr(sv, "CHECKPOINT_ATTRS"):
+        snapshot[sk] = copy.copy(sv)
+        for kk in sv.CHECKPOINT_ATTRS:
+          attr = getattr(sv, kk)
+          copied_attr = jax.tree_map(
+              lambda x: x.copy() if hasattr(x, 'copy') else x, attr)
+          setattr(sv, kk, copied_attr)
```

Comment:
@michaeloneill I am not sure I 100% understand what you intend to do in your suggestion because `snapshot_state` does not exist. Moreover, I do not think one gets around to copying the `experiment` object (`copy.copy(sv)` in my suggestion). If the experiment is not copied, its variables, including e.g. the network parameters, will change over time. I think this should also affect the non-async modes with the in-memory checkpointing of jaxline (i.e. not only `train_eval_multithreaded`). Or what am I missing?

Also note that I used the "duck-typing" as this seems to be the default throughout `jaxline.utils`. E.g. see https://github.com/google-deepmind/jaxline/blob/53a2bc97c25dfa981528e493394833da429461d9/jaxline/utils.py#L544

What seems to work is:
```
    # Ensure buffers do not get donated as training loop runs ahead.
    for k, v in snapshot.items():
      if k == 'experiment_module':
        snapshot[k] = copy.copy(v)
        snapshot_state = jax.tree_map(_maybe_device_get, v.snapshot_state())
        snapshot[k].restore_from_snapshot(snapshot_state)
```

Reply:
snapshot_state is a method on the AbstractExperiment class, which your experiment presumably inherits from. That method implicitly copies the checkpointable attributes via get_first, so it should be all you need. 

Yes either one could be used for duck typing, but I think checking for 'experiment_module' is cleaner. 

Could you try with my suggested edit and make sure this does indeed work for you?